tex/generic/pgf/frontendlayer/tikz/tikz.code.tex

```diff
@@ -2123,125 +2123,42 @@
-    \else%
-      \ifx\pgf@let@token g%
-        \let\pgfutil@next=\tikz@g@char%
-      \else%
-        \ifx\pgf@let@token s%
-           \let\pgfutil@next=\tikz@schar%
-        \else%
-          \ifx\pgf@let@token |%
-             \let\pgfutil@next=\tikz@vh@lineto%
-          \else%
-            \ifx\pgf@let@token p%
-              \let\pgfutil@next=\tikz@pchar%
-              \pgfsetmovetofirstplotpoint%
-            \else%
-              \ifx\pgf@let@token t%
-                \let\pgfutil@next=\tikz@to%
-              \else%
-                \ifx\pgf@let@token\pgfextra%
-                  \let\pgfutil@next=\tikz@extra%
-                \else%
-                  \ifx\pgf@let@token\foreach%
-                    \let\pgfutil@next=\tikz@foreach%
-                  \else%
-                    \ifx\pgf@let@token f%
-                      \let\pgfutil@next=\tikz@fchar%
-                    \else%
-                      \ifx\pgf@let@token\pgf@stop%
-                        \let\pgfutil@next=\relax%
-                      \else%
-                        \ifx\pgf@let@token\par%
-                          \let\pgfutil@next=\tikz@scan@next@command%
-                        \else%
-                          \ifx\pgf@let@token d%
-                            \let\pgfutil@next=\tikz@decoration%
-                          \else%
-                            \ifx\pgf@let@token l%
-                              \let\pgfutil@next=\tikz@l@char%
-                            \else%
-                              \ifx\pgf@let@token:%
-                                \let\pgfutil@next=\tikz@colon@char%
-                              \else%
-                                \let\pgfutil@next=\tikz@expand%
-                              \fi%
-                            \fi%
-                          \fi%
-                        \fi%
-                      \fi%
-                    \fi%
-                  \fi%
-                \fi%
-              \fi%
-            \fi%
-          \fi%
-        \fi%
-      \fi%
-    \fi%
-  \fi%
-  \ifx\pgfutil@next\tikz@expand\else\tikz@expandcount=100\relax\fi%
-  \pgfutil@next%
+  \pgfutil@switch\pgfutil@ifx\pgf@let@token{%
+    {(}{\tikz@movetoabs}%)
+    {+}{\tikz@movetorel}%
+    {-}{\tikz@lineto}%
+    {.}{\tikz@dot}%
+    {r}{\tikz@rect}%
+    {n}{\tikz@fig}%
+    {[}{\tikz@parse@options}%]
+    {c}{\tikz@cchar}%
+    {\bgroup}{\tikz@beginscope}%
+    {\egroup}{\tikz@endscope}%
+    {;}{\tikz@finish}%
+    {a}{\tikz@a@char}%
+    {e}{\tikz@e@char}%
+    {g}{\tikz@g@char}%
+    {s}{\tikz@schar}%
+    {|}{\tikz@vh@lineto}%
+    {p}{\pgfsetmovetofirstplotpoint\tikz@pchar}%
+    {t}{\tikz@to}%
+    {\pgfextra}{\tikz@extra}%
+    {\foreach}{\tikz@foreach}%
+    {f}{\tikz@fchar}%
+    {\pgf@stop}{\relax}%
+    {\par}{\tikz@scan@next@command}%
+    {d}{\tikz@decoration}%
+    {l}{\tikz@l@char}%
+    {:}{\tikz@colon@char}%
+    {\relax}{\relax\tikz@scan@next@command}%
+  }{}{\tikz@resetexpandcount\tikz@expand}%
```

Comment:
This is buggy. The count should be reset in case of match, rather than not match. Current code will result in endless loop.

Moving `\tikz@resetexpandcount` to `#4` of `\pgfutil@switch` doesn't work either because then e.g. `\tikz@movetoabs\tikz@resetexpandcount` is left in input stream, but we need them in reversed order, at least the expansion order.

Reply:
A possible fix https://github.com/muzimuzhi/pgf/compare/ea44e24e5f721b9b4a00e26f8291ebfc76522ff6...6ffd8f92cbd84eeaaadee161f5d7a59fc9e000a3, note this makes the semantic of `\pgfutil@switch` un-conventional.

Reply:
I think we have to go with
```latex
  % ..
  {:}{\let\pgf@next\tikz@colon@char}%
  {\relax}{\relax\let\pgf@next\tikz@scan@next@command}%
}{\tikz@resetexpandcount\pgf@next}{\tikz@expand}%
```
Sorry, but I want to avoid the changed semantics of `\pgfutil@switch`. In fact I was thinking about moving this to a separate `pgfexpl` module which can serve as a drop-in replacement for `expl3` in case we want to switch to that in the future.  That would make migration as easy as doing a simple search-and-replace.

Reply:
That's rather reasonable (and exciting for the `pgfexpl` part).  Could I understand it as you have assigned yourself to the fixing and new `pgfexpl` tasks? 😄

Reply:
Ah, I can do the part of fixing uses of `\pgfutil@switch`. And leave the part of moving and renaming `\pgfutil@switch` (and possibly `pgfint.code.tex`) to you.

Reply:
Found it's `\token_case_meaning:Nn` that `\pgfutil@switch\pgfutil@ifx` imitates. I should read some `expl3` codes ...